pynisher/support.py

```diff
@@ -0,0 +1,158 @@
+
+
+@lru_cache(maxsize=None)
+def _has_pywin32() -> bool:
+    """Check if system has pywin32 accessible"""
+    try:
+        import win32api  # noqa
+
+        return True
+    except Exception:
+        return False
+
+
+def supports(limit: str) -> bool:
+    """Check if pynisher supports a given feature
+
+    Parameters
+    ----------
+    limit: "walltime" | "cputime" | "memory" | "decorator"
+        The kind of feature to check support for
+
+    Returns
+    -------
+    bool
+        Whether it is supported or not
+    """
+    mapping = {
+        "walltime": supports_walltime,
+        "cputime": supports_cputime,
+        "memory": supports_memory,
+        "decorator": supports_limit_decorator,
+    }
+    func = mapping.get(limit.lower(), None)
+    if func is not None:
+        return func()
+    else:
+        raise ValueError(f"Not a known feature, must be one of {list(mapping.keys())}")
+
+
+def supports_walltime() -> bool:
+    """Check if wall time is supported on this system
+
+    * Linux - Yes
+    * Darwin - Yes
+    * Windows - Yes
+
+    Check respective "pynisher/limiters/<platform>.py"
+
+    Returns
+    -------
+    bool
+        Whether it's supported or not
+    """
+    plat = sys.platform.lower()
+    if plat.startswith("linux"):
+        # Sure, should work
+        return True
+    elif plat.startswith("darwin"):
+        # Yup, also works
+        return True
+    elif plat.startswith("win"):
+        # Yay, they all work
+        return True
+    else:
+        raise NotImplementedError(f"Unknown system {platform.platform()}")
+
+
+def supports_cputime() -> bool:
+    """Check if cpu time is supported on this system
+
+    * Linux - Yes
+    * Darwin - Yes
+    * Windows - Yes
+
+    Check respective "pynisher/limiters/<platform>.py"
+
+    Returns
+    -------
+    bool
+        Whether it's supported or not
+    """
+    plat = sys.platform.lower()
+    if plat.startswith("linux"):
+        # Sure, should work
+        return True
+    elif plat.startswith("darwin"):
+        # Yup, also works
+        return True
+    elif plat.startswith("win"):
+        # Yup, we can do so with pywin32
+        return _has_pywin32()
+    else:
+        raise NotImplementedError(f"Unknown system {platform.platform()}")
+
+
+def supports_memory() -> bool:
+    """Check if memory limit is supported on this system
+
+    * Linux - Yes
+    * Darwin - No
+    * Windows - If `pywin32` installed correctly
+
+    Check respective "pynisher/limiters/<platform>.py"
+
+    Returns
+    -------
+    bool
+        Whether it's supported or not
+    """
+    plat = sys.platform.lower()
+    if plat.startswith("linux"):
+        # Sure, should work
+        return True
+
+    elif plat.startswith("darwin"):
+        # Can't limit memory on Mac
+        return True
+
+    elif plat.startswith("win"):
+        # Only supported if we can successfuly import `pywin32` modules
+        return _has_pywin32()
```

Comment:
What happens if `pywin32` is installed and cannot be imported? Wouldn't it make sense to also give a warning then?

Reply:
The `_has_pywin32` check does exactly that, we care if it can be used.

https://github.com/automl/pynisher/blob/477714f41e829f278cd68d070b36d2c853472695/pynisher/support.py#L13-L21

If you ignore the `pynisher.supports('cpu_time')` ability to check, then trying to limit cputime will raise an error.

https://github.com/automl/pynisher/blob/477714f41e829f278cd68d070b36d2c853472695/pynisher/limiters/windows.py#L141-L144

Reply:
I.e. This methodology extends to Mac and Windows (when pywin32) is not setup correctly.

```python
if not pynisher.supports("memory") and memory_limit is not None:
    # ... warn or error?
    memory_limit = None
```